### PR TITLE
Add user profile form

### PR DESCRIPTION
**Pull request by jwalgran**
## Overview

Adds a second form shown after logging in for the first time that prompts for additional, optional information.

Connects #2255 

### Demo

screen shot 2017-10-25 at 2 38 16 pm

### Notes

Following the pattern of `ItsiUser` I have created a new model related to `User` to hold additional, optional information.

- All additional user profile fields are optional.
- `user_type` is selected from a defined list of choices and defaults to "Unspecified."
- `country` is selected from a defined list of choices and defaults to "US."

Based on the style of the login modal form, I created a new user profile modal form

- The form is shown after logging in only if both the `was_skipped` and `is_complete` flags on the `UserProfile` are false.
- Making the decision whether or not to show the form required passing the login response as an argument through the chain of success event handlers.
- The options for the `country` and `user_type` fields are passed from Python to Javascript via the existing `window.clientSettings` object.
- The nunjucks template is built at compile time, not at run time, so there is some Javascript that completes the construction of the option tags on the client.
- I edited `bundle.sh` to include the `bootstrap-select` stylesheet that was previously added to the project.

## Testing Instructions

 * Run migrations and rebuild the vendor bundle.
 * Log in and verify that the new form is displayed.
 * Click "I'll do this later." Log out, log back in, and verify that the new form is not shown.
 * Log in as a different user.
 * Fill out the form and click "Submit profile"
 * Log out, log back in, and verify that the new form is not shown.
 * Run './scripts/manage.sh dbshell` and verify the profile data
```sql
select * from user_userprofile;
```

### CODE REVIEW

**Comment by jwalgran** [comment]
Fixing lint now

File: src/mmw/apps/user/models.py [hunk 1]
```diff
@@ -32,3 +34,41 @@ class ItsiUser(models.Model):
...
+        (UNI_UGRAD_STU, UNI_UGRAD_STU),
+        (COMMERCIAL, COMMERCIAL),
+        (GOVERNMENT, GOVERNMENT),
+        (K_12_STU, K_12_STU),
+        (K_12_TEACHER, K_12_TEACHER),
+        (OTHER, OTHER),
+    )
+
+    user = models.OneToOneField(User, primary_key=True)
+    was_skipped = models.BooleanField(default=False)
```
Comment (arottersman) [inline src/mmw/apps/user/models.py]: Will this always be the negation of `is_complete`?

Comment (jwalgran) [reply src/mmw/apps/user/models.py]: For now, yes, but here is the workflow where they will not be:

- Log in for the first time
- Click "...later...". `was_skipped` is set to true.
- Using the feature in #2256 fill out the form. `is_complete` is set to true.

**Comment by ajrobbins** [comment]
The styling on the new modal doesn't match the wireframes, but we should change all the styling at once in https://github.com/WikiWatershed/model-my-watershed/issues/2254

**Comment by ajrobbins** [comment]
>Pure keyboard entry does not seem to be possible because of bootstrap select, which forces the use of a mouse. 

Chatted with Terence and ideally the selection would work as follows:
- Tabbing into the dropdown opens it
- typing while in the dropdown should scroll to the matching value (or at least the first letter)
- hitting Enter from within the dropdown should select the selected value and close the dropdown, not submit the whole form

**Comment by jwalgran** [comment]
I am seeing the keyboard navigation work as described in Arianna's previous comment. In this GIF I did not touch the mouse.

mmw-profile-keyboard-nav

The support is there, but I suspect we may be running into a bug.

It looks like the problem that prompted us to fork bootstrap-select may have been fixed in a more recent release. I will try updating that dependency and see if it makes the behavior of bootstrap-select more stable.

**Comment by rajadain** [comment]
An issue was that if I tried to hit <kbd>Enter</kbd> or <kbd>Esc</kbd> it would submit / dismiss the modal, whereas I just wanted to close the dropdown or select the highlighted element. If we disable those actions while focused on the select, that may be enough.

**Comment by jwalgran** [comment]
In the GIF I am using <kbd>Enter</kbd> to select the highlighted item. If I press enter twice, it submits the form. This is what I would expect. Pressing enter in a text field submits the form, too.

The <kbd>Esc</kbd> behavior is not ideal, though. I would expect it to behave like <kbd>Enter</kbd>. First press on the key closes the dropdown. Second press ends up dismissing the form in the same way that pressing <kbd>Esc</kbd> in a text box dismisses the form.

File: src/mmw/apps/user/views.py [hunk 2]
```diff
@@ -41,12 +41,15 @@ def login(request):
...
                 auth_login(request, user)
+                profile, created = UserProfile.objects.get_or_create(user=user)
                 response_data = {
                     'result': 'success',
                     'username': user.username,
                     'itsi': ItsiUser.objects.filter(user_id=user.id).exists(),
                     'guest': False,
-                    'id': user.id
+                    'id': user.id,
+                    'profileWasSkipped': profile.was_skipped,
```
[the remark this answers is not in the document]
Comment (jwalgran) [reply src/mmw/apps/user/views.py]: Addressed in fixup commit c134926

File: src/mmw/apps/user/views.py [hunk 3]
```diff
@@ -41,12 +41,15 @@ def login(request):
...
+                    'profileIsComplete': profile.is_complete
```
[the remark this answers is not in the document]
Comment (jwalgran) [reply src/mmw/apps/user/views.py]: Addressed in fixup commit c134926

File: src/mmw/js/src/app.js [hunk 4]
```diff
@@ -138,10 +138,36 @@ var App = new Marionette.Application({
         });
     },
 
+
     showLoginModal: function(onSuccess) {
+        var self = this;
+        function promptForProfileIfIncomplete(loginResponse) {
```
[the remark this answers is not in the document]
Comment (jwalgran) [reply src/mmw/js/src/app.js]: Addressed in fixup commit c134926

File: src/mmw/js/src/user/views.js [hunk 5]
```diff
@@ -260,6 +261,68 @@ var LoginModalView = ModalBaseView.extend({
...
+        this.$el.modal('hide');
+    },
+
+    handleFailure: function(response) {
+        this.handleServerFailure(response);
+    },
+
+    setFields: function () {
+        this.model.set({
+            first_name: $(this.ui.firstName.selector).val(),
```
[the remark this answers is not in the document]
Comment (jwalgran) [reply src/mmw/js/src/user/views.js]: Addressed in fixup commit c134926

**Comment by jwalgran** [comment]
@rajadain I added code in 517083f to make the esc key behave the way we want it to.

mmw-esc-key-in-dropdown

**Comment by rajadain** [comment]
Taking another look.

**Comment by rajadain** [comment]
When I press Enter once on the select menu to open it, it still submits the form, as shown [here](https://i.imgur.com/HoE78Ob.gifv).

Given that we would like this functionality in the November release, but there is plenty of budget in the project, I think we should defer this polish until later. We can also chat on Monday morning where I can show you this behavior on my screen in case that may help.

**Comment by jwalgran** [comment]
In fixup commit 547a51d I added a handler to suppress the enter key event on dropdowns. It is a little strange, but it fixed the problem for me.

**Comment by rajadain** [comment]
Taking a final look

**Comment by rajadain** [comment]
Hmm, I'm not seeing the options in the dropdown anymore:

image

**Comment by rajadain** [comment]
Sorry, that was my fault. Hadn't bundled the scripts correctly.

+1 tested, it's working now, pressing Enter doesn't accidentally submit the form, and I was able to fully complete it using the keyboard!

**Comment by rajadain** [comment]
Friendly reminder to squash the fixups before merging.

**Pull request merged**